google/api_core/retry_async.py

```diff
@@ -74,23 +94,23 @@ async def check_if_exists():
 async def retry_target(
     target, predicate, sleep_generator, timeout=None, on_error=None, **kwargs
 ):
-    """Call a function and retry if it fails.
+    """Await a coroutine and retry if it fails.
```

Comment:
IIUC, the sleep-delay logic should be the same between all four versions of this function: syncvs async, unary vs streaming. I would suggest:

* at the very least making the four functions look as similar as possible with their timeout handling logic. For example, the async versions uses `min()`, which the sync version probably should as well. The streaming functions use `exception_factory`; should the unary ones as well, for consistenct?

* consider whether it's possible and desirable to factor any of the time- and exception-handling functionality into a common function. I suspect the answer is "no" because it would make the code more complex and harder to follow, but it would have the advantage that we'd be using the same logic in the parts where it makes sense. Maybe just factor out the `if deadline` and `_LOGGER` at the bottom of the loop into a helper function that all four versions call to increment the sleep consistently?

Reply:
I was trying to avoid making changes to the existing retries, but I suppose with the new base classes, it makes sense to generalize them a bit. I added a shared `_retry_error_helper` function, which either raises an error, or logs the retry message.

Note though that I had to make a (potentially breaking) change to the async unary retry logic to make it consistent though. Previously async_unary would attempt to cancel early when the deadline is reached, while the other retries only check the deadline after the request terminates. The old behavior was very slow (using asyncio.wait_for), and caused [race conditions](https://github.com/googleapis/python-api-core/issues/528). I'd consier it a bug, but fixing the behavior may count as a breaking change (even though gapic functions shouldn't be affected)